For EL rebuilds, users often want EPEL this permits a clean way to add it

jcpunk opened the pull request:
Hello,

From the Scientific Linux side, we'd like to add EPEL to the list of disabled additional software repos.

It shouldn't be enabled by default, but it is something a lot of our users utilize.

This code adds a place within the InstallClass were distribution owners can add 3rd party repos they believe their community may desire.

This may also be helpful for pre-seeding the CentOS folks.  They could use this to add some of their most popular SIGs to the list.

jkonecny12 commented on pyanaconda/payload/__init__.py:
```diff
@@ -989,9 +988,30 @@ def __init__(self, data):
+            if not self.isRepoEnabled(repo.name):
+                log.debug("Removing disabled repo %s", repo.name)
+                self.removeRepo(repo.name)
+
```
Why we need to remove disabled repos? They should be ignored for installation anyway.

jcpunk replied:
They are ignored by the installer, but pykickstart sees them in the list and adds them to anaconda-ks.cfg (even though they are disabled)

jkonecny12 commented on pyanaconda/payload/dnfpayload.py:
```diff
@@ -380,35 +382,39 @@ def _add_repo(self, ksrepo):
         else:
             with self._repos_lock:
                 self._base.repos.add(repo)
-            repo.enable()
```
Why these two `repo.enable()` are not here anymore?

jcpunk replied:
I consolidated them into a single repo.enable after the log message below.

jcpunk replied:
Pondering a bit more, I shifted the fetch into a second function so I can make a clean `addDisabledRepo` that re-uses most of `_add_repo`

jkonecny12 commented on pyanaconda/payload/dnfpayload.py:
```diff
@@ -380,35 +382,39 @@ def _add_repo(self, ksrepo):
             if not url and not mirrorlist and not metalink:
-                self._base.repos[repo.id].enable()
+                if enabled:
+                    self._base.repos[repo.id].enable()
```
Isn't this repo enable useless when you will call it anyway in the 
```
if enabled:
    self._base.repos[repo.id].enable()
```
a few lines below?

jcpunk replied:
I believe it probably is, but I was not sure how best to document the behaviour being captured by the if statements above.  I'm open to ideas there.

jkonecny12 commented:
Jenkins, it's ok to test.

jcpunk commented:
Alas, I can't see the failed jenkins errors.....

jkonecny12 commented:
Yeah, sorry for that. I have 'make test results public' on my TODO list but I'm swamped by other priority work.

Here important parts of the test results:
```

======================================================================
ERROR: Failure: SyntaxError (invalid syntax (installclass_test.py, line 190))
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/usr/lib/python3.6/site-packages/nose/loader.py", line 417, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/lib/python3.6/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/lib/python3.6/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/usr/lib64/python3.6/imp.py", line 235, in load_module
    return load_source(name, filename, file)
  File "/usr/lib64/python3.6/imp.py", line 172, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 674, in exec_module
  File "<frozen importlib._bootstrap_external>", line 781, in get_code
  File "<frozen importlib._bootstrap_external>", line 741, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/anaconda/tests/pyanaconda_tests/installclass_test.py", line 190
    self.assertTrue(hasattr(testclass, 'hidden')
       ^
SyntaxError: invalid syntax

----------------------------------------------------------------------
Ran 343 tests in 15.417s

FAILED (SKIP=1, errors=1, failures=1)
FAIL nosetests.sh (exit status: 1)

FAIL: pylint/runpylint.py
=========================

************* Module ../pyanaconda/payload/dnfpayload.py
E1123(unexpected-keyword-arg):1112,12: DNFPayload.updateBaseRepo: Unexpected keyword argument 'enable' in method call
************* Module ../pyanaconda/installclasses/scientific.py
E0001(syntax-error): 53,0: : unexpected EOF while parsing (<string>, line 53)
************* Module ../tests/pyanaconda_tests/installclass_test.py
E0001(syntax-error):190,0: : invalid syntax (<string>, line 190)
FAIL pylint/runpylint.py (exit status: 2)

SKIP: storage/run_storage_tests.py
==================================
```

and broken build:
```
/usr/lib/rpm/sepdebugcrcfix: Updated 5 CRC32s, 0 CRC32s did match.
611 blocks
+ /usr/lib/rpm/check-buildroot
+ /usr/lib/rpm/redhat/brp-ldconfig
/sbin/ldconfig: Warning: ignoring configuration file that cannot be opened: /builddir/build/BUILDROOT/anaconda-28.20-1.fc28.x86_64/etc/ld.so.conf: No such file or directory
+ /usr/lib/rpm/brp-compress
+ /usr/lib/rpm/brp-strip-static-archive /usr/bin/strip
+ /usr/lib/rpm/brp-python-bytecompile /usr/bin/python 1
Bytecompiling .py files below /builddir/build/BUILDROOT/anaconda-28.20-1.fc28.x86_64/usr/lib64/python3.6 using /usr/bin/python3.6
*** Error compiling '/builddir/build/BUILDROOT/anaconda-28.20-1.fc28.x86_64/usr/lib64/python3.6/site-packages/pyanaconda/installclasses/scientific.py'...
  File "/usr/lib64/python3.6/scientific.py", line 52
    metalink="https://mirrors.fedoraproject.org/metalink?repo=epel-"+major_version+"&arch="+productArch)
                                                                                                       ^
SyntaxError: unexpected EOF while parsing
RPM build errors:
error: Bad exit status from /var/tmp/rpm-tmp.agjeJ2 (%install)
    Bad exit status from /var/tmp/rpm-tmp.agjeJ2 (%install)
Child return code was: 1
EXCEPTION: [Error()]
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/mockbuild/trace_decorator.py", line 96, in trace
    result = func(*args, **kw)
  File "/usr/lib/python3.6/site-packages/mockbuild/util.py", line 618, in do
    raise exception.Error("Command failed: \n # %s\n%s" % (command, output), child.returncode)
mockbuild.exception.Error: Command failed: 
```

Unit tests will fail even when you fix those, we have one failing test on Rawhide which I'm trying to fix right now.

jcpunk commented:
I'll get those sorted and squash down to one commit.

jkonecny12 commented:
jenkins, test this please

jkonecny12 commented:
This is the last bit.
```
************* Module ../pyanaconda/payload/dnfpayload.py
E1123(unexpected-keyword-arg):1115,12: DNFPayload.updateBaseRepo: Unexpected keyword argument 'enable' in method call
FAIL pylint/runpylint.py (exit status: 2)
```

jcpunk commented:
Should be fixed now

jkonecny12 commented:
Yes, sorry for delay. We have plenty of other work but I will try to look on that tomorrow.

jcpunk commented:
I'll keep an eye out.  I broke the test for the non-new behaviors out into a separate PR so that doesn't get lost in the shuffle.

jkonecny12 commented:
PR https://github.com/rhinstaller/anaconda/pull/1353 was merged so you can start work on this when you are ready.

jkonecny12 commented:
BTW stupid question, but do you want this to F28? If yes, I must rebase that PR on f28-devel branch.

jcpunk commented:
Awesome, I'll take a look.

My ultimate target is the RHEL8 branch, but I'm planning to start with anaconda/master and then some backport PRs to Fedora/RHEL8.

jcpunk commented:
In theory, I've got a viable candidate pushed up just now.

jkonecny12 commented on pyanaconda/payload/__init__.py:
```diff
@@ -992,6 +1012,13 @@ def __init__(self, data):
+        for repo in self.disabledRepos:
+            log.debug("Removing disabled repo %s", repo.name)
+            self.removeRepo(repo.name)
+
```
Interesting I didn't know this is happening and basically you are correct it doesn't give sense to have disabled repo in an output kickstart file. 
However, you should remove the `TODO` and fix it in pyanaconda/kickstart.py the enabled part is not part of kickstart project but Anaconda's extension. Basically you need to add the `__str__()` method to the `RepoData` class and exclude disabled repositories there.

jcpunk replied:
Ahh, that's where that's happening.  I'll get it fixed.

jkonecny12 commented on pyanaconda/payload/dnfpayload.py:
```diff
@@ -327,11 +327,13 @@ def _replace_vars(self, url):
         return None
 
-    def _add_repo(self, ksrepo):
+    def _add_repo(self, ksrepo, enabled=True):
```
Please remove these changes you have done in the `dnfpayload` file completely. They are too invasive. Instead do the changes only in the `updateBaseRepo` method.

jcpunk replied:
With removing the flag for `_add_repo` I end up duplicating a lot of the code in that function.  Additionally my quick tests show the `_add_repo()` inside of `updateBaseRepo` ends up with the repo enabled, rather than disabled.

Not sure what you'd like to see there.  `_add_repo()` is private to `dnfpayload.py`

jcpunk commented on pyanaconda/payload/__init__.py:
```diff
@@ -460,10 +480,10 @@ def updateBaseRepo(self, fallback=True, checkmount=True):
-        system.  The repo will be automatically enabled and its metadata
-        fetched.
+        system.
+        If enabled, the repo metadata will be fetched automatically.
```
Can do, I'll have this done up shortly.

jcpunk commented:
Just checking back in.

I've not been able to find a use case for
```
        for ksrepo in self.data.repo.dataList():
            log.debug("repo %s: mirrorlist %s, baseurl %s, metalink %s",
                      ksrepo.name, ksrepo.mirrorlist, ksrepo.baseurl, ksrepo.metalink)
            # one of these must be set to create new repo
            if not (ksrepo.mirrorlist or ksrepo.baseurl or ksrepo.metalink or
                    ksrepo.name in self._base.repos):
                raise payload.PayloadSetupError("Repository %s has no mirror, baseurl or metalink set "
                                                "and is not one of the pre-defined repositories"
                                                % ksrepo.name)
            self._add_repo(ksrepo)
```
from dnfpayload.py

I did some test installs on F28 without it and had no issues.

If I can pull that block out, anaconda doesn't wrongly enable my `configurePayload` repos.

jkonecny12 commented:
That part of the code is fixing this: https://bugzilla.redhat.com/show_bug.cgi?id=1215963 so it shouldn't be removed.

jkonecny12 commented:
Please also fix those:
```
************* Module ../pyanaconda/payload/__init__.py
W0311(bad-indentation):397,0: : Bad indentation. Found 11 spaces, expected 12
W0311(bad-indentation):398,0: : Bad indentation. Found 15 spaces, expected 16
W0311(bad-indentation):407,0: : Bad indentation. Found 11 spaces, expected 12
W0311(bad-indentation):408,0: : Bad indentation. Found 15 spaces, expected 16
************* Module ../pyanaconda/installclasses/scientific.py
E1121(too-many-function-args): 56,8: ScientificBaseInstallClass.configurePayload: Too many positional arguments for method call
************* Module ../tests/pyanaconda_tests/payload_test.py
W0102(dangerous-default-value): 88,4: DummyRepo.__init__: Dangerous default value [] as argument
```

And sorry that these results are not public. I really need to implement a way how to export them outside.

jcpunk commented:
I think I've got that mostly sorted.... I'm not seeing what is wrong with `super(ScientificBaseInstallClass, self).configurePayload(self, payload)`

Not sure why my local pylint check didn't kick those back.....

I read up on https://bugzilla.redhat.com/show_bug.cgi?id=1215963

The exception checking all makes sense it is the `self._add_repo(ksrepo)` that is biting me.

jcpunk commented:
.... Found it, no need for `self` in the args list using super.

jkonecny12 commented on pyanaconda/kickstart.py:
```diff
@@ -1814,6 +1814,13 @@ def __init__(self, *args, **kwargs):
+    def __str__(self):
+        """Don't output disabled repos"""
+        if self.enabled:
+            return super(RepoData, self).__str__(self)
```
I'm getting exception thanks to this line. Please change ti to `super().__str__()`.

jcpunk commented:
I think I'm closer now.  I still end up with the disabled repos enabled by `updateBaseRepo`....

I noticed `ksnames` in `updateBaseRepo` isn't used by anything so I dropped it.

Also, switched `self.data.repo.dataList()` to use the payload property of `self.addOns`.

jcpunk commented:
I think I've got it all fixed finally.

jcpunk commented:
Just checking back in on this

jkonecny12 commented:
Sorry for delay. I'm swamped by other work right now, I hope I'll be able to look on this soon.

jcpunk commented:
rebased off master

jkonecny12 commented:
This is the reason for test failure:
```
************* Module ../pyanaconda/payload/dnfpayload.py
E0602(undefined-variable):1155,20: DNFPayload.updateBaseRepo: Undefined variable '_fetch_md'
FAIL pylint/runpylint.py (exit status: 2)
```

jcpunk commented:
Right, fixed.

jkonecny12 commented on pyanaconda/payload/dnfpayload.py:
```diff
@@ -381,35 +381,58 @@ def _add_repo(self, ksrepo):
+        :returns: None
+        """
+        ksrepo.disable()
+        self._add_repo(ksrepo)
```
I think you should call `super().addDisabledRepo(ksrepo)` after this line. Otherwise parent will enable this repo back.

jcpunk replied:
good call!

jkonecny12 commented on pyanaconda/payload/dnfpayload.py:
```diff
@@ -1109,16 +1138,22 @@ def updateBaseRepo(self, fallback=True, checkmount=True):
+            enabled_repos = self.enabledRepos
+            for repo in self.addOns:
+                if repo.name in enabled_repos:
+                    self._fetch_md(repo)
```
Couldn't this fetch the metadata second time for some repos? I guess it is not such a problem but still.

jcpunk replied:
That is a bit of a puzzle....

jcpunk commented:
In theory I've fixed up and rebased off master

jcpunk commented:
I'm not sure why my test installs are passing but jenkins shows them failing....  I've gotta be doing something wrong...

jcpunk commented:
In theory this latest push once again works.

jkonecny12 commented:
It is little more complicated. I've just changed test installs so that is the reason I guess.

jkonecny12 commented on pyanaconda/payload/__init__.py:
```diff
@@ -389,6 +389,26 @@ def mirrorEnabled(self):
+        """A list of enabled repos."""
+        enabled = []
+        for repo in self.addOns:
+            if self.isRepoEnabled(repo.name):
```
This line has still the issue.
You can't call `repo.name` from the `self.addOns` output because it returns string not objects.

jcpunk commented:
Fixed

jcpunk commented:
Rebased
